Mathlib/AlgebraicGeometry/Morphisms/Constructors.lean

```diff
@@ -242,6 +251,142 @@ lemma MorphismProperty.topologically_propertyIsLocalAtTarget
+    targetAffineLocally (of P) = P := by
+  ext X Y f
+  constructor
+  · intro hf
+    simp only [targetAffineLocally, Subtype.forall] at hf
+    let 𝒰 : Y.OpenCover := Y.affineCover
+    apply hP.of_openCover f 𝒰
+    intro i
+    have hiao : IsAffineOpen (Scheme.Hom.opensRange (𝒰.map i)) :=
+      AlgebraicGeometry.isAffineOpen_opensRange _
+    rw [← hP.RespectsIso.arrow_mk_iso_iff <| morphismRestrictOpensRange f (𝒰.map i)]
+    exact hf (Scheme.Hom.opensRange (𝒰.map i)) hiao
+  · intro hf ⟨U, hU⟩
+    exact hP.restrict _ _ hf
+
+/-- The restriction of a morphism property of schemes that is local at the target to morphisms
+with affine target, is local. -/
+lemma of_isLocal_of_isLocalAtTarget (P : MorphismProperty Scheme)
+    (hP : PropertyIsLocalAtTarget P) : (of P).IsLocal where
+  RespectsIso := by
+    apply AffineTargetMorphismProperty.respectsIso_mk
+    · intro X Y Z e f _ hf
+      apply hP.RespectsIso.left e f hf
+    · intro X Y Z e f _ hf
+      apply hP.RespectsIso.right e f hf
+  toBasicOpen {X Y} _ f hf := hP.restrict f _
+  ofBasicOpenCover {X Y} _ f s hs hf := by
+    apply ((hP.openCover_TFAE f).out 0 5).mpr
+    let U (r : s) : Opens Y.carrier := Y.basicOpen r.val
+    have hiao : IsAffineOpen (⊤ : Opens Y.carrier) := isAffineOpen_top Y
+    have hU : iSup U = ⊤ := by
+      erw [hiao.basicOpen_union_eq_self_iff]
+      exact hs
+    use s, U, hU, hf
+
+lemma stableUnderBaseChange_of_stableUnderBaseChangeOnAffine_of_isLocalAtTarget
+    (P : MorphismProperty Scheme) (hP₁ : PropertyIsLocalAtTarget P)
+    (hP₂ : (of P).StableUnderBaseChange) :
+    P.StableUnderBaseChange := by
+  rw [← targetAffineLocally_of_eq_of_isLocalAtTarget P hP₁]
+  apply (of_isLocal_of_isLocalAtTarget P hP₁).stableUnderBaseChange
+  exact hP₂
+
+end AffineTargetMorphismProperty
+
+namespace MorphismProperty
+
+open AffineTargetMorphismProperty
+
+/-- A morphism property of schemes is `StableUnderAffineBaseChange` if
+the base change along a morphism of affine schemes of a morphism with affine target is affine. -/
+def StableUnderAffineBaseChange (P : MorphismProperty Scheme) : Prop :=
+  ∀ {X Y Y' S : Scheme} [IsAffine X] [IsAffine S] (f : X ⟶ S) (g : Y ⟶ S) (f' : Y' ⟶ Y)
+    (g' : Y' ⟶ X) (_ : IsPullback f' g' g f) (_ : P g), P g'
+
+lemma StableUnderAffineBaseChange.mk {P : MorphismProperty Scheme}
+    (hP₁ : P.RespectsIso)
+    (hP₂ : ∀ (X Y S : Scheme) [IsAffine X] [IsAffine S] (f : X ⟶ S) (g : Y ⟶ S),
+      P g → P (pullback.fst : pullback f g ⟶ X)) :
+    StableUnderAffineBaseChange P := fun {X Y Y' S} _ _ f g f' g' sq hg ↦ by
+  let e := sq.flip.isoPullback
+  rw [← hP₁.cancel_left_isIso e.inv, sq.flip.isoPullback_inv_fst]
+  exact hP₂ _ _ _ f g hg
+
+/-- If `P` is local at the target, to show that `P` is stable under base change, it suffices to
+check this for base change along a morphism of affine schemes. -/
+lemma stableUnderBaseChange_of_isLocalAtTarget_of_stableUnderAffineBaseChange
```

Comment:
This could be called `StableUnderAffineBaseChange.stableUnderBaseChange`

Reply:
Changed.